kasa/device_factory.py

```diff
@@ -1,123 +1,83 @@
-        )
-        if protocol_class is not None:
-            dev.protocol = protocol_class(
-                host,
-                transport=AesTransport(
-                    host, port=port, credentials=credentials, timeout=timeout
-                ),
-            )
-        await dev.update()
+    def _perf_log(has_params, perf_type):
+        nonlocal start_time
         if debug_enabled:
             end_time = time.perf_counter()
             _LOGGER.debug(
-                "Device %s with known type (%s) took %.2f seconds to connect",
-                host,
-                device_type.value,
-                end_time - start_time,
+                f"Device {cparams.host} with connection params {has_params} "
+                + f"took {end_time - start_time:.2f} seconds to {perf_type}",
             )
-        return dev
-
-    unknown_dev = SmartDevice(
-        host=host, port=port, credentials=credentials, timeout=timeout
-    )
-    if protocol_class is not None:
-        # TODO this will be replaced with connection params
-        unknown_dev.protocol = protocol_class(
-            host,
-            transport=AesTransport(
-                host, port=port, credentials=credentials, timeout=timeout
-            ),
+            start_time = time.perf_counter()
+
+    if (protocol := get_protocol(cparams=cparams)) is None:
+        raise UnsupportedDeviceException(
+            f"Unsupported device for {cparams.host}: "
+            + f"{cparams.connection_type.device_family.value}"
         )
-    await unknown_dev.update()
-    device_class = get_device_class_from_sys_info(unknown_dev.internal_state)
-    dev = device_class(host=host, port=port, credentials=credentials, timeout=timeout)
-    # Reuse the connection from the unknown device
-    # so we don't have to reconnect
-    dev.protocol = unknown_dev.protocol
-    await dev.update()
-    if debug_enabled:
-        end_time = time.perf_counter()
-        _LOGGER.debug(
-            "Device %s with unknown type (%s) took %.2f seconds to connect",
-            host,
-            dev.device_type.value,
-            end_time - start_time,
+
+    device_class: Optional[Type[SmartDevice]]
+
+    if isinstance(protocol, TPLinkSmartHomeProtocol):
+        info = await protocol.query(GET_SYSINFO_QUERY)
+        _perf_log(True, "get_sysinfo")
+        device_class = get_device_class_from_sys_info(info)
+        device = device_class(cparams.host, port=cparams.port, timeout=cparams.timeout)
+        device.update_from_discover_info(info)
+        device.protocol = protocol
+        await device.update()
+        _perf_log(True, "update")
+        return device
+    elif device_class := get_device_class_from_family(
+        cparams.connection_type.device_family.value
+    ):
+        device = device_class(
+            cparams.host,
+            port=cparams.port,
+            timeout=cparams.timeout,
+            credentials=cparams.credentials,
+        )
```

Comment:
The reason for the factories is to avoid circular dependencies.  Because the device construction needs to reference device classes but the smartdevice class is referencing the connection parameters in `Connect()`.

Reply:
I think `connect()` on smartdevice class should simply pass the incoming parameters to the factory, which should avoid this exact problem, no?